src/middleware/error.rs

```diff
@@ -18,6 +20,18 @@ pub enum MiddlewareInnerError {
     MaxLength(String, usize, usize),
     #[error("{0} amount of {1} should be {1} but it's {2}")]
     DiffAmount(String, String, usize, usize),
+    #[error("{0} should be assigned the value {1} but has previously been assigned {2}")]
+    InvalidWildcardAssignment(Wildcard, Value, Value),
+    #[error("{0} does not match {1}")]
+    MismatchedAnchoredKeyInStatementTmplArg(Key, Key),
+    #[error("{0} does not match against {1}")]
+    MismatchedStatementTmplArg(StatementTmplArg, StatementArg),
+    #[error("{0} does not match {1}")]
+    MismatchedWildcardValueAndStatementArg(Value, Value),
+    #[error("Not all statement templates of {0} have been matched")]
+    UnsatisfiedCustomPredicateConjunction(CustomPredicate),
+    #[error("None of the statement templates of {0} have been matched")]
+    UnsatisfiedCustomPredicateDisjunction(CustomPredicate),
```

Comment:
The `Display` of a `CustomPredicate` has many new lines: https://github.com/0xPARC/pod2/blob/2c41a6c5543a3167c816d6e668478c3cd6f13700/src/middleware/custom.rs#L326-L357

I would suggest rewriting the error message so that the `{0}` appears at the end after a new line like:
`"Not all statement templates of the following custom predicate have been matched:\n{0}"`